src/org/apache/xpath/functions/FuncFilter.java

```diff
@@ -0,0 +1,237 @@
+ * fn:filter is one of XPath 3.1's higher-order function
+ * (ref, https://www.w3.org/TR/xpath-functions-31/#higher-order-functions).
+ * 
+ * The XPath function fn:filter has following signature, and definition,
+ * 
+ * fn:filter($seq as item()*, $f as function(item()) as xs:boolean) as item()*
+ * 
+ * The function fn:filter returns those items from the sequence $seq for which the 
+ * supplied function $f returns true.
+ * 
+ * Error conditions,
+   As a consequence of the function signature and the function calling rules, a type 
+   error occurs if the supplied function $f returns anything other than a single 
+   xs:boolean item. There is no conversion to an effective boolean value.
+ */
+public class FuncFilter extends Function2Args {
+
+   private static final long serialVersionUID = 2912594883291006421L;
+   
+   private static final String FUNCTION_NAME = "filter()"; 
+
+  /**
+   * Execute the function. The function must return a valid object.
+   * 
+   * @param xctxt The current execution context.
+   * 
+   * @return A valid XObject.
+   *
+   * @throws javax.xml.transform.TransformerException
+   */
+  public XObject execute(XPathContext xctxt) throws javax.xml.transform.TransformerException
+  {
+      
+        XObject funcEvaluationResult = null;
+        
+        final int contextNode = xctxt.getCurrentNode();
+        
+        Expression arg0 = getArg0();
+        Expression arg1 = getArg1();                
+        
+        DTMIterator arg0DtmIterator = null;
+        
+        XObject arg0XsObject = null;
+                  
+        if (arg0 instanceof LocPathIterator) {
+            arg0DtmIterator = arg0.asIterator(xctxt, contextNode);               
+        }
+        else {
+            arg0XsObject = arg0.execute(xctxt, contextNode);
+        }
+        
+        ResultSequence resultSeq = new ResultSequence();
+                    
+        if (arg1 instanceof InlineFunction) {            
+            resultSeq = evaluateFnFilter(xctxt, arg0XsObject, arg0DtmIterator, (InlineFunction)arg1); 
+        }
+        else if (arg1 instanceof Variable) {
+            XObject arg1VarValue = arg1.execute(xctxt);
+            if (arg1VarValue instanceof InlineFunction) {
+                resultSeq = evaluateFnFilter(xctxt, arg0XsObject, arg0DtmIterator, (InlineFunction)arg1VarValue);   
+            }
+            else {
+                throw new javax.xml.transform.TransformerException("FORG0006 : The second argument to function call filter(), "
+                                                                                               + "is not a function item.", xctxt.getSAXLocator());    
+            }
+        }
+        else {
+            throw new javax.xml.transform.TransformerException("FORG0006 : The second argument to function call filter(), "
+                                                                                           + "is not a function item.", xctxt.getSAXLocator());               
```

Comment:
It is bad when different errors have exactly the same text. It would be better if the error messages were different, and it would help if they included something related to the actual type of the argument